File: LeetCode/Medium/1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void inOrder(TreeNode* root, vector<int>& v) {
        if (!root) {
            return;
        }
        inOrder(root->left, v);
        v.push_back(root->val);
        inOrder(root->right, v);
    }
    TreeNode* buildTree(vector<int>& v, int start, int end) {
        if (start > end)
            return NULL;

        int mid = start + (end - start) / 2;
        // cout << mid << v[mid] << endl;
        TreeNode* root = new TreeNode(v[mid]);
        root->left = buildTree(v, start, mid - 1);
        root->right = buildTree(v, mid + 1, end);
        return root;
    }
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> nodes;
        inOrder(root, nodes);
        // sort(nodes.begin(), nodes.end());
        TreeNode* newRoot = buildTree(nodes, 0, nodes.size() - 1);
        return newRoot;
    }
};

File: LeetCode/Medium/1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func inorder(root *TreeNode, nodes *[]int) {
	if root == nil {
		return
	}

	inorder(root.Left, nodes)
	*nodes = append(*nodes, root.Val)
	inorder(root.Right, nodes)
}

func buildBalanced(nodes []int, start, end int) (root *TreeNode) {
	if start > end {
		return nil
	}

	mid := (start + end) / 2
    val := nodes[mid]
	newRoot := &TreeNode{Val: val}
	newRoot.Left = buildBalanced(nodes, start, mid-1)
	newRoot.Right = buildBalanced(nodes, mid+1, end)
	return newRoot
}

func balanceBST(root *TreeNode) *TreeNode {
	nodes := []int{}
	inorder(root, &nodes)

	return buildBalanced(nodes, 0, len(nodes)-1)
}